File: 349_IntersectionOfTwoArrays.cpp
//53.67%
class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res;
		
		unordered_map<int, int> unique;
		for(int num : nums1)
			unique[num] = 0;
		for(int num : nums2){
			if(unique.find(num) != unique.end()){
				if(unique[num] == 0){
					res.push_back(num);
					unique[num] = 1;
				}
			}
		}
		return res;
    }
};

//Seems more clean than mine
class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        unordered_set<int> m(nums1.begin(), nums1.end());
        vector<int> res;
        for (auto a : nums2)
            if (m.count(a)) {
                res.push_back(a);
                m.erase(a);
            }
        return res;
    }
};

//A solution which used more 'C++' thinking
vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
    set<int> s(nums1.begin(), nums1.end());
    vector<int> out;
    for (int x : nums2)
        if (s.erase(x))
            out.push_back(x);
    return out;
}

File: 349_IntersectionOfTwoArrays.py
# 0.94%
class Solution(object):
    def intersection(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        return list(set([num for num in nums2 if num in set(nums1)]))

# Better solution than mine		
class Solution(object):
def intersection(self, nums1, nums2):
    """
    :type nums1: List[int]
    :type nums2: List[int]
    :rtype: List[int]
    """
    nums1=set(nums1)
    nums2=set(nums2)
    return list(nums1&nums2)